src/net_processing.cpp

```diff
@@ -1613,8 +1611,12 @@ void static ProcessGetData(CNode* pfrom, const CChainParams& chainparams, CConnm
             if (pfrom->fPauseSend)
                 break;
 
-            const CInv &inv = *it;
-            it++;
+            const CInv &inv = *it++;
+
+            if (pfrom->m_tx_relay == nullptr) {
+                // Ignore GETDATA requests for transactions from blocks-only peers.
+                continue;
```

Comment:
Seems mildly wasteful to grab the `cs_main` lock just to skip over `MSG_TX` and `MSG_WITNESS_TX`. I think this patch ("ignore tx GETDATA from blocks only peers") could be dropped entirely, allowing the "ignore unknown INV types in GETDATA messages" patch's handling to cover it instead.

Reply:
Yeah, it's _slightly_ wasteful, but I think it's clearer to have tx and non-tx handling fully separated in different parts of the functions, and explicitly have a code path/comment documenting behaviour, rather than having to think "so what happens if we receive a request for a tx from a blocks-relay-only peer? Oh, it falls through here and then we handle it below as if it were an unknown type".